File: solutions/go/parsing-log-files/1/parsing_log_files.go
package parsinglogfiles

import (
	"fmt"
	"regexp"
)

func IsValidLine(text string) bool {
	re := regexp.MustCompile(`^\[(TRC|DBG|INF|WRN|ERR|FTL)\]`)

	return re.MatchString(text)
}

func SplitLogLine(text string) []string {
	re := regexp.MustCompile(`<[-=~*]*>`)

	return re.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	re := regexp.MustCompile(`(?i)".*password.*"`)

	count := 0

	for _, line := range lines {
		if re.MatchString(line) {
			count++
		}
	}

	return count
}

func RemoveEndOfLineText(text string) string {
	re := regexp.MustCompile(`end-of-line\d*`)

	return re.ReplaceAllString(text, "")
}

func TagWithUserName(lines []string) []string {
	re := regexp.MustCompile(`User\s+(\w+)`)
	newLines := []string{}

	for _, line := range lines {
		found := re.FindStringSubmatch(line)

		if found != nil {
			line = fmt.Sprintf("[USR] %s %s", found[1], line)
		}

		newLines = append(newLines, line)
	}

	return newLines
}

File: solutions/csharp/parsing-log-files/1/ParsingLogFiles.cs
using System.Text.RegularExpressions;

public class LogParser
{
    private const string IsValidLinePattern = @"^\[(TRC|DBG|INF|WRN|ERR|FTL)\]";
    private const string LogLinePattern = @"<.*?>";
    private const string QuotedPasswordsPattern = @""".*password.*""";
    private const string EndOfLinePattern = @"end-of-line\d*";
    private const string PasswordsPattern = @"password\w+";

    public bool IsValidLine(string text) => new Regex(IsValidLinePattern).IsMatch(text);

    public string[] SplitLogLine(string text) => new Regex(LogLinePattern).Split(text);

    public int CountQuotedPasswords(string lines) => new Regex(QuotedPasswordsPattern, RegexOptions.IgnoreCase).Matches(lines).Count;

    public string RemoveEndOfLineText(string line) => new Regex(EndOfLinePattern).Replace(line, "");

    public string[] ListLinesWithPasswords(string[] lines)
    {
        var newLines = new List<string>();

        foreach (var line in lines)
        {
            var match = new Regex(PasswordsPattern, RegexOptions.IgnoreCase).Match(line);

            if (match.Success)
            {
                newLines.Add($"{match.Value}: {line}" );
            }
            else {
                newLines.Add($"--------: {line}");
            }
        }

        return [.. newLines];
    } 

}
